src/loki/reach.cc

```diff
@@ -110,17 +131,117 @@ directed_reach SimpleReach(const DirectedEdge* edge,
+  directed_reach reach{};
+  if (costing->GetEdgeFilter()(edge) == 0) {
+    return reach;
+  }
+
+  // fake up the input location
+  const baldr::GraphTile* tile = nullptr;
+  const auto* node = reader.GetEndNode(edge, tile);
+  auto ll = node->latlng(tile->header()->base_ll());
+  locations_.Mutable(0)->mutable_ll()->set_lng(ll.first);
+  locations_.Mutable(0)->mutable_ll()->set_lat(ll.second);
+  locations_.Mutable(0)->mutable_path_edges(0)->set_graph_id(edge_id);
+  locations_.Mutable(0)->mutable_path_edges(0)->mutable_ll()->set_lng(ll.first);
+  locations_.Mutable(0)->mutable_path_edges(0)->mutable_ll()->set_lat(ll.second);
+
+  // fake up the costing array
+  std::shared_ptr<sif::DynamicCost> costings[static_cast<int>(sif::TravelMode::kMaxTravelMode)];
+  costings[static_cast<int>(costing->travel_mode())] = costing;
+
+  // expand in the forward direction
+  if (direction | kOutbound) {
+    Clear();
+    Compute(locations_, reader, costings, costing->travel_mode());
+    reach.outbound = std::min(static_cast<uint32_t>(done_.size() - transitions_), max_reach);
+  }
+
+  // expand in the reverse direction
+  if (direction | kInbound) {
+    Clear();
+    ComputeReverse(locations_, reader, costings, costing->travel_mode());
+    reach.inbound = std::min(static_cast<uint32_t>(done_.size() - transitions_), max_reach);
+  }
+
+  return reach;
+}
+
+// callback fired when a node is expanded from, the node will be the end node of the previous label
+void Reach::ExpandingNode(baldr::GraphReader& graphreader,
+                          const baldr::GraphTile* tile,
+                          const baldr::NodeInfo* node,
+                          const sif::EdgeLabel& current,
+                          const sif::EdgeLabel* previous) {
+  // compute the nodes id
+  GraphId node_id = tile->header()->graphid();
+  node_id.set_id(node - tile->node(0));
+  // mark the node and if its a new one then mark its doppelgänger on the other levels
+  if (done_.insert(node_id).second) {
+    for (const auto& trans : tile->GetNodeTransitions(node)) {
+      done_.insert(trans.endnode());
+      ++transitions_;
+    }
+  }
+}
+
+// when the main loop is looking to continue expanding we tell it to terminate here
+thor::ExpansionRecommendation Reach::ShouldExpand(baldr::GraphReader& graphreader,
+                                                  const sif::EdgeLabel& pred,
+                                                  const thor::InfoRoutingType route_type) {
+  if (done_.size() < max_reach_)
+    return thor::ExpansionRecommendation::continue_expansion;
+  return thor::ExpansionRecommendation::prune_expansion;
+}
+
+// tell the expansion how many labels to expect and how many buckets to use
+void Reach::GetExpansionHints(uint32_t& bucket_count, uint32_t& edge_label_reservation) const {
+  // TODO: tweak these for performance
+  bucket_count = max_reach_ * 50;
+  edge_label_reservation = max_reach_ * 10;
+}
+
+void Reach::Clear() {
```

Comment:
this resets for another forward or backward run of either conservative or exact reach